go/tools/builders/nogo_change.go

```diff
@@ -0,0 +1,325 @@
+type DiagnosticEntry struct {
+	analysis.Diagnostic
+	*analysis.Analyzer
+}
+
+// This file contains two main entities: Edit and Change, which correspond to the low-level
+// and high-level abstractions. See them below.
+
+// The following is about the `Edit`, a low-level abstraction of edits.
+// An Edit describes the replacement of a portion of a text file.
+type Edit struct {
+	New   string `json:"new"`   // the replacement
+	Start int    `json:"start"` // starting byte offset of the region to replace
+	End   int    `json:"end"`   // (exclusive) ending byte offset of the region to replace
+}
+
+func (e Edit) String() string {
+	return fmt.Sprintf("{Start:%d,End:%d,New:%q}", e.Start, e.End, e.New)
+}
+
+// SortEdits orders a slice of Edits by (start, end) offset.
+// This ordering puts insertions (end = start) before deletions
+// (end > start) at the same point, but uses a stable sort to preserve
+// the order of multiple insertions at the same point.
+// (Apply detects multiple deletions at the same point as an error.)
+func SortEdits(edits []Edit) {
+	sort.Stable(editsSort(edits))
+}
+
+type editsSort []Edit
+
+func (a editsSort) Len() int { return len(a) }
+func (a editsSort) Less(i, j int) bool {
+	if cmp := a[i].Start - a[j].Start; cmp != 0 {
+		return cmp < 0
+	}
+	return a[i].End < a[j].End
+}
+func (a editsSort) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// UniqueEdits returns a list of edits that is sorted and
+// contains no duplicate edits. Returns the index of some
+// overlapping adjacent edits if there is one and <0 if the
+// edits are valid.
+// Deduplication helps in the cases where two analyzers produce duplicate edits.
+func UniqueEdits(edits []Edit) ([]Edit, int) {
+	if len(edits) == 0 {
+		return nil, -1
+	}
+	equivalent := func(x, y Edit) bool {
+		return x.Start == y.Start && x.End == y.End && x.New == y.New
+	}
+	SortEdits(edits)
+	unique := []Edit{edits[0]}
+	invalid := -1
+	for i := 1; i < len(edits); i++ {
+		prev, cur := edits[i-1], edits[i]
+		if !equivalent(prev, cur) {
+			unique = append(unique, cur)
+			if prev.End > cur.Start {
+				invalid = i
+			}
+		}
+	}
+	return unique, invalid
+}
+
+// ApplyEdits applies a sequence of edits to the src buffer and returns the
+// result. Edits are applied in order of start offset; edits with the
+// same start offset are applied in they order they were provided.
+//
+// ApplyEdits returns an error if any edit is out of bounds,
+// or if any pair of edits is overlapping.
+func ApplyEdits(src string, edits []Edit) (string, error) {
+	edits, size, err := validate(src, edits)
+	if err != nil {
+		return "", err
+	}
+
+	// Apply edits.
+	out := make([]byte, 0, size)
+	lastEnd := 0
+	for _, edit := range edits {
+		if lastEnd < edit.Start {
+			out = append(out, src[lastEnd:edit.Start]...)
+		}
+		out = append(out, edit.New...)
+		lastEnd = edit.End
+	}
+	out = append(out, src[lastEnd:]...)
+
+	if len(out) != size {
+		panic("wrong size")
+	}
+
+	return string(out), nil
```

Comment:
can we return bytes instead? The `out` is converted to string here and immediately converted back to []byte by its only caller

Reply:
done